numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -373,31 +374,62 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
-
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    static PyObject *fallback = NULL;
+    npy_cache_import("numpy.core.arrayprint", "_scalar_fallback", &fallback);
+    if (fallback == NULL) {
+        return NULL;
     }
-    return ret;
+    return PyObject_CallFunction(fallback, "O", self);
 }
 
 
 static PyObject *
 gentype_repr(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
+    PyObject  *dtype_type, *dtype_name, *item, *item_repr;
 
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        /* XXX: Why are we using str here? */
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    dtype_type = (PyObject*)PyArray_DescrFromScalar(self)->typeobj;
+    dtype_name = PyObject_GetAttrString(dtype_type, "__name__");
+    if (dtype_name == NULL) {
+        /* this should never happen since types always have __name__ */
+        PyErr_Format(PyExc_RuntimeError, "missing __name__ attribute");
+        return NULL;
     }
-    return ret;
+
+    item = gentype_generic_method(self, NULL, NULL, "item");
+    if (item == NULL) {
+        Py_DECREF(dtype_name);
+        return NULL;
+    }
+
+    item_repr = PyObject_Repr(item);
+    Py_DECREF(item);
+    if (item_repr == NULL) {
+        Py_DECREF(dtype_name);
+        return NULL;
+    }
+
+    PyUString_ConcatAndDel(&dtype_name, PyUString_FromString("("));
+    PyUString_ConcatAndDel(&dtype_name, item_repr);
+    PyUString_ConcatAndDel(&dtype_name, PyUString_FromString(")"));
+    return dtype_name;
 }
 
+static PyObject *
+genint_type_str(PyObject *self)
+{
+    PyObject  *item, *item_repr;
+    item = gentype_generic_method(self, NULL, NULL, "item");
+    if (item == NULL) {
+        return NULL;
+    }
+
+    item_repr = PyObject_Str(item);
+    Py_DECREF(item);
+    return item_repr;
+}
```

Comment:
I guess our intention is slightly more explicit (doesn't go to the 'last resort'), and it's a slight optimization since we avoid calling `get_format_function`.

I think that with this only void (and user-defined) types use the scalar fallback.